tests/test_clean.py

```diff
@@ -1,96 +1,210 @@
+    (
+        'text<!-- comment -->',
+        True,
+        'text'
+    ),
+    (
+        'text<!--comment-->',
+        True,
+        'text'
+    ),
+    (
+        'text<!-- comment -->',
+        False,
+        'text<!-- comment -->'
+    ),
+    (
+        'text<!--comment-->',
+        False,
+        'text<!--comment-->'
     )
+])
+def test_comments(data, should_strip, expected):
+    assert clean(data, strip_comments=should_strip) == expected
 
 
-def test_named_arguments():
-    ATTRS = {'a': ['rel', 'href']}
+@pytest.mark.parametrize('data, expected', [
+    # Disallowed tag is escaped
+    ('<img src="javascript:alert(\'XSS\');">', '&lt;img src="javascript:alert(\'XSS\');"&gt;'),
+
+    # Test with parens
+    ('a <script>safe()</script> test', 'a &lt;script&gt;safe()&lt;/script&gt; test'),
+
+    # Test with braces
+    ('a <style>body{}</style> test', 'a &lt;style&gt;body{}&lt;/style&gt; test'),
+])
+def test_disallowed_tags(data, expected):
+    assert clean(data) == expected
 
-    text = '<a href="http://xx.com" rel="alternate">xx.com</a>'
-    assert bleach.clean(text) == '<a href="http://xx.com">xx.com</a>'
+
+def test_invalid_char_in_tag():
+    # NOTE(willkg): Two possible outcomes because attrs aren't ordered
+    assert (
+        clean('<script/xss src="http://xx.com/xss.js"></script>') in
+        [
+            '&lt;script src="http://xx.com/xss.js" xss=""&gt;&lt;/script&gt;',
+            '&lt;script xss="" src="http://xx.com/xss.js"&gt;&lt;/script&gt;'
+        ]
+    )
     assert (
-        bleach.clean(text, attributes=ATTRS) ==
-        '<a href="http://xx.com" rel="alternate">xx.com</a>'
+        clean('<script/src="http://xx.com/xss.js"></script>') ==
+        '&lt;script src="http://xx.com/xss.js"&gt;&lt;/script&gt;'
     )
 
 
-def test_disallowed_html():
+def test_unclosed_tag():
+    assert (
+        clean('a <em>fixed tag') ==
+        'a <em>fixed tag</em>'
+    )
     assert (
-        bleach.clean('a <script>safe()</script> test') ==
-        'a &lt;script&gt;safe()&lt;/script&gt; test'
+        clean('<script src=http://xx.com/xss.js<b>') ==
+        '&lt;script src="http://xx.com/xss.js&lt;b"&gt;&lt;/script&gt;'
     )
+    # NOTE(willkg): Two possible outcomes because attrs aren't ordered
     assert (
-        bleach.clean('a <style>body{}</style> test') ==
-        'a &lt;style&gt;body{}&lt;/style&gt; test'
+        clean('<script src="http://xx.com/xss.js"<b>') in
+        [
+            '&lt;script src="http://xx.com/xss.js" &lt;b=""&gt;&lt;/script&gt;',
+            '&lt;script &lt;b="" src="http://xx.com/xss.js"&gt;&lt;/script&gt;'
+        ]
+    )
+    # NOTE(willkg): Two possible outcomes because attrs aren't ordered
+    assert (
+        clean('<script src="http://xx.com/xss.js" <b>') in
+        [
+            '&lt;script src="http://xx.com/xss.js" &lt;b=""&gt;&lt;/script&gt;',
+            '&lt;script &lt;b="" src="http://xx.com/xss.js"&gt;&lt;/script&gt;'
+        ]
```

Comment:
The diff in github here is terrible. I merged a couple of tests between `test_clean.py` and `test_security.py`, but otherwise this is untouched.

Reply:
👍